src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/Whitelist.java

```diff
@@ -66,6 +66,70 @@
 
     public abstract boolean permitsStaticFieldSet(@Nonnull Field field, @CheckForNull Object value);
 
+    // Utility methods for creating canonical string representations of the signature
+    public static final StringBuilder joinWithSpaces(StringBuilder b, String[] types) {
+        for (String type : types) {
+            b.append(' ').append(type);
+        }
+        return b;
+    }
+
+    public static @Nonnull String getName(@Nonnull Class<?> c) {
+        Class<?> e = c.getComponentType();
+        if (e == null) {
+            return c.getName();
+        } else {
+            return getName(e) + "[]";
+        }
+    }
+
+    public static @Nonnull String getName(@CheckForNull Object o) {
+        return o == null ? "null" : getName(o.getClass());
+    }
+
+    public static String[] argumentTypes(Class<?>[] argumentTypes) {
+        String[] s = new String[argumentTypes.length];
+        for (int i = 0; i < argumentTypes.length; i++) {
+            s[i] = getName(argumentTypes[i]);
+        }
+        return s;
+    }
+
+    /** Canonical name for a field access. */
+    public static String canonicalFieldString(@Nonnull Field field) {
+        return getName(field.getDeclaringClass()) + ' ' + field.getName();
+    }
+
+    /** Canonical name for a method call. */
+    public static String canonicalMethodString(@Nonnull Method method) {
+        return joinWithSpaces(new StringBuilder(getName(method.getDeclaringClass())).append(' ').append(method.getName()), argumentTypes(method.getParameterTypes())).toString();
```

Comment:
No, this is **not safe**. You must encode the defining `ClassLoader` somehow. Ditto for all the other key calculations.

Reply:
Could be fixed by using a double-level `WeakHashMap<ClassLoader, Map<String, Boolean>>` I think, since the values would not refer to the keys.

Reply:
or a `WeakHashMap<Class, String>` if you just omit the `Class.name` from the start of the `String` part

Reply:
@jglick This does not make sense -- the Cacheable whitelists are using fixed string signatures -- https://github.com/jenkinsci/script-security-plugin/blob/8a4fa975bc182e3a365db1f9d53b20c3c62015c9/src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist#L94 -- those signatures do not include a ClassLoader, nor can they.

Reply:
Basically: if the static whitelists don't generate a hit, then we inspect the dynamic ones which can look at ClassLoader, etc